File: ts/project/.sdk/tm/php/utility/PrepareQuery.php
<?php
declare(strict_types=1);

// ProjectName SDK utility: prepare_query

class ProjectNamePrepareQuery
{
    public static function call(ProjectNameContext $ctx): array
    {
        $point = $ctx->point;
        $reqmatch = $ctx->reqmatch ?? [];
        $params = [];
        if ($point) {
            $p = \Voxgig\Struct\Struct::getprop($point, 'params');
            if (is_array($p)) {
                $params = $p;
            }
        }
        $out = [];
        $items = \Voxgig\Struct\Struct::items($reqmatch);
        if ($items) {
            foreach ($items as $item) {
                $key = $item[0];
                $val = $item[1];
                if ($val !== null && is_string($key) && !in_array($key, $params, true)) {
                    $out[$key] = $val;
                }
            }
        }
        return $out;
    }
}

File: ts/project/.sdk/tm/java/utility/PrepareQuery.java
package JAVAPACKAGE.utility;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import JAVAPACKAGE.core.Context;
import JAVAPACKAGE.utility.struct.Struct;

@SuppressWarnings({"unchecked"})
final class PrepareQuery {

  private PrepareQuery() {}

  static Map<String, Object> prepareQuery(Context ctx) {
    Map<String, Object> point = ctx.point;
    Map<String, Object> reqmatch = ctx.reqmatch;
    if (reqmatch == null) {
      reqmatch = new LinkedHashMap<>();
    }

    List<Object> params = null;
    if (point != null) {
      Object p = Struct.getprop(point, "params");
      if (p instanceof List) {
        params = (List<Object>) p;
      }
    }
    if (params == null) {
      params = new ArrayList<>();
    }

    Map<String, Object> out = new LinkedHashMap<>();
    for (List<Object> item : Struct.items(reqmatch)) {
      String key = item.get(0) instanceof String ? (String) item.get(0) : "";
      Object val = item.get(1);
      if (val != null && !containsStr(params, key)) {
        out.put(key, val);
      }
    }

    return out;
  }

  private static boolean containsStr(List<Object> list, String s) {
    for (Object v : list) {
      if (v instanceof String && v.equals(s)) {
        return true;
      }
    }
    return false;
  }
}
